lib/options.go

```diff
@@ -44,6 +46,117 @@ const DefaultScenarioName = "default"
 
+// DNSConfig is the DNS resolver configuration.
+type DNSConfig struct {
+	// If positive, defines how long DNS lookups should be returned from the cache.
+	TTL null.String `json:"ttl"`
+	// Strategy to use when picking a single IP if more than one is returned for a host name.
+	Strategy NullDNSStrategy `json:"strategy"`
+	// FIXME: Valid is unused and is only added to satisfy some logic in
+	// lib.Options.ForEachSpecified(), otherwise it would panic with
+	// `reflect: call of reflect.Value.Bool on zero Value`.
+	Valid bool
+}
+
+// DNSStrategy is the strategy to use when picking a single IP if more than one
+// is returned for a host name.
+//go:generate enumer -type=DNSStrategy -transform=kebab -trimprefix DNS -output dns_strategy_gen.go
+type DNSStrategy uint8
+
+// NullDNSStrategy is a nullable wrapper around DNSStrategy, required for the
+// current configuration system.
+type NullDNSStrategy struct {
+	DNSStrategy
+	Valid bool
+}
+
+const (
+	// DNSFirst returns the first IP from the response.
+	DNSFirst DNSStrategy = iota + 1
+	// DNSRoundRobin rotates the IP returned on each lookup.
+	DNSRoundRobin
+	// DNSRandom returns a random IP from the response.
+	DNSRandom
+)
+
+// DefaultDNSConfig returns the default DNS configuration.
+func DefaultDNSConfig() DNSConfig {
+	return DNSConfig{
+		TTL:      null.NewString("5m", false),
+		Strategy: NullDNSStrategy{DNSRandom, false},
+	}
+}
+
+// String implements fmt.Stringer.
+func (c DNSConfig) String() string {
+	out := make([]string, 0, 2)
+	out = append(out, fmt.Sprintf("ttl=%s", c.TTL.String))
+	out = append(out, fmt.Sprintf("strategy=%s", c.Strategy.String()))
+	return strings.Join(out, ",")
+}
+
+// MarshalJSON implements json.Marshaler.
+func (c DNSConfig) MarshalJSON() ([]byte, error) {
+	var s string
+	if c.Strategy.IsADNSStrategy() {
+		s = c.Strategy.String()
+	}
+	return json.Marshal(struct {
+		TTL      string `json:"ttl"`
+		Strategy string `json:"strategy"`
+	}{
+		TTL:      c.TTL.String,
+		Strategy: s,
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (c *DNSConfig) UnmarshalJSON(data []byte) error {
+	params := make(map[string]interface{})
+	if err := json.Unmarshal(data, &params); err != nil {
+		return err
+	}
+	return c.unmarshal(params)
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (c *DNSConfig) UnmarshalText(text []byte) error {
+	if string(text) == DefaultDNSConfig().String() {
+		*c = DefaultDNSConfig()
+		return nil
+	}
+	params, err := strvals.Parse(string(text))
+	if err != nil {
+		return err
+	}
+	return c.unmarshal(params)
+}
+
+func (c *DNSConfig) unmarshal(params map[string]interface{}) error {
+	for k, v := range params {
+		switch k {
+		case "strategy":
+			if v == "" {
+				continue
+			}
+			if s, err := DNSStrategyString(v.(string)); err != nil {
+				return err
+			} else { // nolint: golint
+				c.Strategy.DNSStrategy = s
+				c.Strategy.Valid = true
+			}
```

Comment:
Since `s` is used in the `else` block, declaring it outside "pollutes" the outer scope, IMO. This is a moot point in this case, but in general I think this is an edge case that `golint` doesn't detect properly. See https://github.com/golang/lint/issues/202 .

But sure, I can switch to that style if you think it's more readable, even though I dislike that it has functional implications in the code and shouldn't be linted...

Reply:
Resolved by 8662aa9c.